README.md
Munin - MySQL
=============

Munin plugin for showing graphs of MySQL resource usage.


Installation
------------

1. Download zip file from Github and unzip (https://github.com/meersjo/munin-mysql/archive/master.zip)
2. Install dependencies 
    - munin-node
    - Perl modules: DBI, DBD::mysql, Module::Pluggable
3. Edit Makefile
4. Edit mysql.conf
5. Run `make install'


Further Information
-------------------

The plugin documentation is contained in the plugin file as POD. View
it with perldoc.

The SchemaSize plugin performs a pretty heavy query on information_schema.tables.
Careful when using this on a large environment. Since it is SLOW, you may also 
need to set timeout in the plugin config. See http://munin-monitoring.org/wiki/plugin-conf.d
On 5.1.17+, there's a DIRTY HACK in the code that speeds it up for InnoDB tables.
Should have zero impact, but you may still want to read the official
documentation on the setting:
http://dev.mysql.com/doc/refman/5.1/en/innodb-parameters.html#sysvar_innodb_stats_on_metadata

I don't have a lower version to test on, but it should simply pass the hack and
work slowly.

There is a blog post with some general information and screenshots at
<http://oierud.name/bliki/ImprovedMuninGraphsForMySQL.html>

The information on
<http://code.google.com/p/mysql-cacti-templates/wiki/MySQLTemplates>
should be relevant (As these munin graphs are heavily inspired by
Xaprb's cacti graphs.)

### Wiki

There is a wiki at <https://github.com/kjellm/munin-mysql/wiki>

### Source code

The source is hosted at github:
<http://github.com/meersjo/munin-mysql/>


Troubleshooting
---------------

- If you get warnings saying "Output from SHOW ENGINE INNODB STATUS
  was truncated" that means that a very large deadlock are causing the
  output to be truncated. The consequence is that data for many of the
  InnoDB related data sources will be missing. For solutions to this
  problem see
  http://www.xaprb.com/blog/2006/08/08/how-to-deliberately-cause-a-deadlock-in-mysql/

- You can find some tips for debugging munin plugin problems here:
  <http://munin.projects.linpro.no/wiki/Debugging_Munin_plugins>

- Bugs should be reported to the issue tracker on github
  <http://github.com/meersjo/munin-mysql/issues>


Authors
-------

Johan De Meersman &lt;vegivamp AT tuxera DOT be&gt;

Blatantly stolen, I mean forked off the code by Kjell-Magne Øierud &lt;kjellm AT oierud DOT net&gt;

Inspired by the cacti graphs made by Xaprb
http://code.google.com/p/mysql-cacti-templates/ as viewed on
http://www.xaprb.com/blog/2008/05/25/screenshots-of-improved-mysql-cacti-templates/.

This plugin also contains changes from the upstream munin mysql plugin. See [the list of contributors](https://github.com/munin-monitoring/munin/blob/devel/authors).



CHANGES
2014-06-02: version 0.4

 - Plenty of various tweaks

 - Added a new SchemaSize module (WARNING: SLOW)

 - Threw around the init sequence and added support for in-module collect_data() sub


2010-02-16: version 0.3.1

 - Changed data source type for Opened_tables to DERIVE (again).

 - Don't overwrite mysql_.conf if it already exists on make install.


2010-02-02: version 0.3

 - Added info to a lot of data sources

 - Changed data source type for Opened_tables to DERIVE

 - Support for monitoring multiple mysql instances

 - Improved compatibility with MySQL 4.1

 - Added graphs: uptime, processlist status

 - Fixed issues with the munin protocol commands autoconf and suggest

 - Fixed the "InnoDB Buffer Pool" graph config.

 - Updated vlabels for consistency


2009-10-30: version 0.2

 - Made the plugin more fit for inclusion by the Munin project. Note
   that this breaks backwards compatibility since some data_sources
   have changed names.

 - Handle new "BACKGROUND THREAD" section in SHOW ENGINE INNODB STATUS.

 - The plugin did not work when InnoDB is disabled.

 - Prevented that handled DBI errors from leaking to error messages.

 - Added proper vlabels to the Query Cache graphs.

 - Added a base and lower limit to the Query Cache Memory graph.

 - Plugin now supports the autoconf capability

 - Fixed infinite loop caused by truncated InnoDB status.


2009-03-21: version 0.1.2

 - Improved tests

 - More robust InnoDB status parsing. Handle different order of the
   sections. Handle 64 bit integers written in hexadecimal notation.

 - Added autoconf capability.
  
 - More documentation
  
 - Renamed a graph: mysql_queries -> mysql_commands. This was done so
   that you can have the original mysql graphs side-by-side with the
   ones provided here.
  
 - Easier installation
  
 - Removed the 'check' configuration option -- it didn't really
   work very well.


2009-02-21: version 0.1.1

 - Fixed bug that made parsing of SHOW ENGINE INNODB STATUS enter an
   infinite loop

 - Removed dependency on Math::BigInt::GMP. Caused segfaults.
 
 - Changed drawing type LINE to LINE1 for better compability with
   older versions of munin/rrdtool


2009-02-13: version 0.1

 - Initial release


lib/Munin/MySQL/Graph/InnoDB.pm
package Munin::MySQL::Graph::InnoDB;

use warnings;
use strict;

sub graphs { return {
    innodb_bpool => {
        config => {
            global_attrs => {
                title  => 'InnoDB Buffer Pool',
                vlabel => 'Pages',
                args   => "--base 1024",
            },
            data_source_attrs => {
                draw => 'LINE1',
                type => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'ib_bpool_size',     label  => 'Buffer pool size',
                                          draw   => 'AREA',
                                          colour => 'ffd660',
                                          info   => 'The total number of buffer pool pages'},
            {name => 'ib_bpool_dbpages',  label  => 'Database pages',
                                          draw   => 'AREA',
                                          colour => 'cdcfc4',
                                          info   => 'The number of used buffer pool pages'},
            {name => 'ib_bpool_free',     label => 'Free pages',
                                          info  => 'The number of unused buffer pool pages'},
            {name => 'ib_bpool_modpages', label => 'Modified pages',
                                          info  => 'The number of "dirty" database pages'},
            {name => 'ib_bpool_oldpages', label => 'Old pages',
                                          info  => 'The number of pages in the old sublist of the buffer pool'},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_bpool_act => {
        config => {
            global_attrs => {
                title  => 'InnoDB Buffer Pool Activity',
                vlabel => 'Activity per ${graph_period}',
                # Makes spikes harder to interpret
                #total  => 'Total',
            },
            data_source_attrs => {
                draw => 'LINE1',
            },
        },
        data_sources => [
            {name => 'ib_bpool_read',    label => 'Pages read',
                                         info  => 'Pages read into the buffer pool from disk.'},
            {name => 'ib_bpool_created', label => 'Pages created',
                                         info  => 'Pages created in the buffer pool without reading the corresponding disk page.'},
            {name => 'ib_bpool_written', label => 'Pages written',
                                         info  => 'Pages written to disk from the buffer pool.'},
            {name => 'ib_bpool_made_young',
                                         label => 'Made young',
                                         info  => 'The number of old pages that were moved to the head of the buffer pool'},
            {name => 'ib_bpool_made_not_young',
                                         label => 'Made not young',
                                         info  => 'The number of pages that have remained in the old sublist without being made new.'},
        ],
    },


    #---------------------------------------------------------------------

    innodb_bpool_internal_breakdown  => {
        config => {
            global_attrs => {
                title  => 'InnoDB Buffer Pool Internal breakdown',
                args   => '--base 1024 --lower-limit 0',
                vlabel => 'bytes',
            },
            data_source_attrs => {
                min  => '0',
                draw => 'AREASTACK',
                type => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'ib_bpool_internal_adaptive_hash_size_const',    label => 'Adaptive Hash const'},
            {name => 'ib_bpool_internal_adaptive_hash_size_var',      label => 'Adaptive Hash var'},
            {name => 'ib_bpool_internal_page_hash_size_total',        label => 'Page Hash'},
            {name => 'ib_bpool_internal_dictionary_cache_size_const', label => 'Dictionary const'},
            {name => 'ib_bpool_internal_dictionary_cache_size_var',   label => 'Dictionary var'},
            {name => 'ib_bpool_internal_file_system_size_const',      label => 'Filesystem const'},
            {name => 'ib_bpool_internal_file_system_size_var',        label => 'Filesystem var'},
            {name => 'ib_bpool_internal_lock_system_size_const',      label => 'Lock system const'},
            {name => 'ib_bpool_internal_lock_system_size_var',        label => 'Lock system var'},
            {name => 'ib_bpool_internal_recovery_system_size_const',  label => 'Recovery system const'},
            {name => 'ib_bpool_internal_recovery_system_size_var',    label => 'Recovery system var'},
        ],
    },

    #---------------------------------------------------------------------

    innodb_checkpoint_age => {
        config => {
            global_attrs => {
                title  => 'InnoDB Checkpoint Age',
                vlabel => 'Bytes',
                args   => "--base 1024 -l 0",
            },
            data_source_attrs => {
                draw  => 'AREA',
                type  => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'innodb_log_size',  label  => 'InnoDB log size',
                                         colour => 'cdcfc4',
                                         info   => 'The size in bytes of InnoDB log space.',
                                         value  => sub {
                                             $_[0]->{innodb_log_file_size} * $_[0]->{innodb_log_files_in_group}
                                         }},
            {name => 'ib_log_chkpt_age', label  => 'Uncheckpointed bytes',
                                         colour => 'ffd660',
                                         info   => 'The age in bytes of InnoDB checkpoint.',
                                         value  => sub {
                                             exists $_[0]->{ib_log_chkpt_age}
                                               ? $_[0]->{ib_log_chkpt_age}
                                               : $_[0]->{ib_log_written} - $_[0]->{ib_log_checkpoint}
                                         }},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_history_length => {
        config => {
            global_attrs => {
                title  => 'InnoDB History List',
                vlabel => 'Transactions',
            },
            data_source_attrs => {
                draw  => 'LINE1',
                type  => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'ib_tnx_hist', label => 'History list length',
                                    info  => 'Number of unpurged transactions in undo space.'},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_lsn => {
        config => {
            global_attrs => {
                title  => 'InnoDB Log Sequence Number',
            },
            data_source_attrs => {
                draw  => 'LINE1',
            },
        },
        data_sources => [
            { name  => 'Innodb_lsn_current',
              label => 'Current',
              info  => 'Log sequence number as shown in the LOG section of the SHOW ENGINE INNODB STATUS output.'},
            { name  => 'Innodb_lsn_flushed',
              label => 'Flushed',
              info  => 'Flushed up to log sequence number as shown in the LOG section of the SHOW ENGINE INNODB '},
            { name  => 'Innodb_lsn_last_checkpoint', 
              label => 'Last checkpoint',
              info  => 'Log sequence number last checkpoint as shown in the LOG section of the SHOW ENGINE INNODB STATUS output.'},
        ],
    },

    
    #---------------------------------------------------------------------
    
    innodb_srv_master_thread => {
        config => {
            global_attrs => {
                title  => 'InnoDB Master Thread',
            },
            data_source_attrs => {
                draw => 'AREASTACK',
            },
        },
        data_sources => [
            {name => 'Innodb_master_thread_active_loops', label => 'Active Loops'},
            {name => 'Innodb_master_thread_idle_loops',   label => 'Idle Loops'},
            {name => 'ib_srv_main_flush_loops',      label => 'Flush Loop'},
            {name => 'ib_srv_main_background_loops', label => 'Background Loop'},
            {name => 'ib_srv_main_flushs_writes',    label => 'Flushes/Writes', draw  => 'LINE1'},
        ],
    },

    #---------------------------------------------------------------------

    innodb_read_views => {
        config => {
            global_attrs => {
                title  => 'InnoDB Read Views',
                args   => '--lower-limit 0',
            },
            data_source_attrs => {
                type  => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'ib_innodb_read_views', label => 'Views'},
        ],
    },


    #---------------------------------------------------------------------

    innodb_descriptors => {
        config => {
            global_attrs => {
                title  => 'InnoDB Descriptors',
                args   => '--lower-limit 0',
            },
            data_source_attrs => {
                type  => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'ib_innodb_descriptors', label => 'Descriptors'},
            {name => 'ib_innodb_descriptors_max', label => 'Max', draw => 'AREA', colour => 'ffd660'},
        ],
    },

    #---------------------------------------------------------------------

    innodb_queries => {
        config => {
            global_attrs => {
                title  => 'InnoDB Engine Queries and Transactions',
                args   => '--lower-limit 0',
            },
            data_source_attrs => {
                type => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'ib_innodb_queries', label => 'Active'},
            {name => 'ib_innodb_transactions_active', label => 'Transactions'},
            {name => 'ib_innodb_rw_transactions_active', label => 'RW Transactions'},
            {name => 'ib_innodb_ro_transactions_active', label => 'RO Transactions'},
            {name => 'ib_innodb_query_queue_len', label => 'Queued'},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_insert_buf => {
        config => {
            global_attrs => {
                title  => 'InnoDB Insert Buffer',
                vlabel => 'Activity per ${graph_period}',
            },
            data_source_attrs => {
                draw => 'LINE1',
            },
        },
        data_sources => [
            {name => 'ib_ibuf_inserts',    label => 'Merge Inserts',
                                           info  => 'These values shows statistics about how many buffer operations InnoDB has done. The ratio of merges to inserts gives a good idea of how efficient the buffer is.'},
            {name => 'ib_ibuf_delete',     label => 'Merge Deletes'},
            {name => 'ib_ibuf_merged_rec', label => 'Merged Records'},
            {name => 'ib_ibuf_merges',     label => 'Merges'},
            {name => 'ib_ibuf_discard_inserts', label => 'Discard Inserts'},
            {name => 'ib_ibuf_discard_delete',  label => 'Discard Deletes'},
        ],
    },
      
    #---------------------------------------------------------------------
    
    innodb_insert_buf_size => {
        config => {
            global_attrs => {
                title  => 'InnoDB Insert Buffer Size',
                vlabel => 'Pages',
                args   => "-l 0",
            },
            data_source_attrs => {
                draw  => 'LINE1',
                type  => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'ib_ibuf_seg_size', label  => 'Segment size',
                                         draw   => 'AREA',
                                         colour => 'cdcfc4',
                                         info   => 'Allocated size of insert buffer segment.'},
            {name => 'ib_ibuf_size',     label  => 'Unmerged pages',
                                         colour => '0022ff',
                                         info   => 'Number of pages containing unmerged records.'},
            {name => 'ib_ibuf_free_len', label  => 'Free pages',
                                         info   => 'Number of pages which are free.'},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_io => {
        config => {
            global_attrs => {
                title  => 'InnoDB IO',
                vlabel => 'IO operations per ${graph_period}',
            },
            data_source_attrs => {
                draw => 'LINE1',
            },
        },
        data_sources => [
            {name => 'ib_io_read',  label => 'File reads',
                                    info  => 'The number of calls to the OS read function.'},
            {name => 'ib_io_write', label => 'File writes',
                                    info  => 'The number of calls to the OS write function.'},
            {name => 'ib_io_log',   label => 'Log writes',
                                    info  => 'The number of calls to the OS write function that is caused by the log subsystem.'},
            {name => 'ib_io_fsync', label => 'File syncs',
                                    info  => 'The number of calls to the OS fsync function.'},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_io_pend => {
        config => {
            global_attrs => {
                title  => 'InnoDB IO Pending',
                vlabel => 'Pending operations',
            },
            data_source_attrs => {
                draw => 'LINE1',
            },
        },
        data_sources => [
            {name => 'ib_iop_log',         label => 'AIO Log'},
            {name => 'ib_iop_sync',        label => 'AIO Sync'},
            {name => 'ib_iop_flush_bpool', label => 'Buf Pool Flush'},
            {name => 'ib_iop_flush_log',   label => 'Log Flushes'},
            {name => 'ib_iop_ibuf_aio',    label => 'Insert Buf AIO Read'},
            {name => 'ib_iop_aioread',     label => 'Normal AIO Reads'},
            {name => 'ib_iop_aiowrite',    label => 'Normal AIO Writes'},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_log => {
        config => {
            global_attrs => {
                title  => 'InnoDB Log',
                vlabel => 'Log activity per ${graph_period}',
            },
            data_source_attrs => {
                draw => 'LINE1',
            },
        },
        data_sources => [
            {name => 'innodb_log_buffer_size', label  => 'Buffer Size',
                                               type   => 'GAUGE',
                                               draw   => 'AREA',
                                               colour => 'fafd9e',
                                               info   => 'The size in bytes of the buffer that InnoDB  uses to write to the log files on disk.'},
            {name => 'ib_log_flush',           label => 'KB Flushed',
                                               info  => 'Number of bytes flushed to the transaction log file.'},
            {name => 'ib_log_written',         label => 'KB Written',
                                               info  => 'Number of bytes written to the transaction log buffer.'},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_rows => {
        config => {
            global_attrs => {
                title  => 'InnoDB Row Operations',
                vlabel => 'Operations per ${graph_period}',
                # Makes spikes harder to interpret
                #total  => 'Total',
            },
            data_source_attrs => {},
        },
        data_sources => [
            {name => 'Innodb_rows_deleted',  label => 'Deletes',
                                             info  => 'The number of rows deleted from InnoDB tables.'},
            {name => 'Innodb_rows_inserted', label => 'Inserts',
                                             info  => 'The number of rows inserted into InnoDB tables.'},
            {name => 'Innodb_rows_read',     label => 'Reads',
                                             info  => 'The number of rows read from InnoDB tables.'},
            {name => 'Innodb_rows_updated',  label => 'Updates',
                                             info  => 'The number of rows updated in InnoDB tables.'},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_semaphores => {
        config => {
            global_attrs => {
                title  => 'InnoDB Semaphores',
                vlabel => 'Semaphores per ${graph_period}',
            },
            data_source_attrs => {
                draw  => 'LINE1',
            },
        },
        data_sources => [
            {name => 'ib_spin_rounds',        label => 'Spin Rounds',
                                              info => 'The number of InnoDB semaphore/mutex spin rounds for the internal sync array.'},
            {name => 'ib_spin_waits',         label => 'Spin Waits',
                                              info => 'The number of InnoDB semaphore/mutex spin waits for the internal sync array.'},
            {name => 'ib_os_waits',           label => 'OS Waits',
                                              info => 'The number of InnoDB semaphore/mutex waits yielded to the OS.'},
            {name => 'ib_rw_shared_rounds',   label => 'RW/S Rounds',
                                              info => 'The number of shared (read) semaphore spin rounds within the InnoDB sync array.'},
            {name => 'ib_rw_shared_waits',    label => 'RW/S Waits',
                                              info => 'The number of shared (read) semaphore spin waits within the InnoDB sync array.'},
            {name => 'ib_rw_shared_os_waits', label => 'RW/S OS Waits',
                                              info => 'The number of shared (read) semaphore waits yielded to the OS by InnoDB.'},
            {name => 'ib_rw_excl_rounds',     label => 'RW/X Rounds',
                                              info => 'The number of exclusive (write) semaphore spin rounds within the InnoDB sync array.'},
            {name => 'ib_rw_excl_waits',      label => 'RW/X Waits',
                                              info => 'The number of exclusive (write) semaphore spin waits within the InnoDB sync array.'},
            {name => 'ib_rw_excl_os_waits',   label => 'RW/X OS Waits',
                                              info => 'The number of exclusive (write) semaphore waits yielded to the OS by InnoDB.'},
        ],
    },
    
    #---------------------------------------------------------------------
    
    innodb_tnx => {
        config => {
            global_attrs => {
                title  => 'InnoDB Transactions',
                vlabel => 'Transactions per ${graph_period}',
            },
            data_source_attrs => {
                draw  => 'LINE1',
            },
        },
        data_sources => [
            {name => 'ib_tnx', label => 'Transactions created',
                               info  => 'Number of transactions created.'},
        ],
    },
}}

1;



lib/Munin/MySQL/Graph/QueryCache.pm
package Munin::MySQL::Graph::QueryCache;

use warnings;
use strict;

sub graphs { return {
    qcache => {
        config => {
            global_attrs => {
                title  => 'Query Cache',
                vlabel => 'Commands per ${graph_period}',
            },
            data_source_attrs => {
                 draw => 'LINE1',
            },
        },
        data_sources => [
            {name => 'Qcache_queries_in_cache', label => 'Queries in cache (K)',
                                                info  => 'The number of queries registered in the query cache.',
                                                cdef  => 'Qcache_queries_in_cache,1024,/',
                                                type  => 'GAUGE'},
            {name => 'Qcache_hits',             label => 'Cache hits',
                                                info  => 'The number of query cache hits.'},
            {name => 'Qcache_inserts',          label => 'Inserts',
                                                info  => 'The number of queries added to the query cache.'},
            {name => 'Qcache_not_cached',       label => 'Not cached',
                                                info  => 'The number of noncached queries (not cacheable, or not cached due to the query_cache_type setting).'},
            {name => 'Qcache_lowmem_prunes',    label => 'Low-memory prunes',
                                                info  => 'The number of queries that were deleted from the query cache because of low memory.'},
        ],
    },

    #---------------------------------------------------------------------

    qcache_mem => {
        config => {
            global_attrs => {
                title  => 'Query Cache Memory',
                vlabel => 'Bytes',
                args   => "--base 1024 --lower-limit 0",
            },
            data_source_attrs => {
                draw => 'AREA',
                type => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'query_cache_size',    label => 'Cache size',
                                            info  => 'The amount of memory allocated for caching query results.'},
            {name => 'Qcache_free_memory',  label => 'Free mem',
                                            info  => 'The amount of free memory for the query cache.'},
        ],
    },

    #---------------------------------------------------------------------

    qcache_avg_block_size=> {
        config => {
            global_attrs => {
                title  => 'Query Cache Average Block Size',
                vlabel => 'Bytes',
                args   => "--base 1024 --lower-limit 0",
            },
            data_source_attrs => {
                draw => 'LINE1',
                type => 'GAUGE',
            },
        },
        data_sources => [
            {name => 'query_cache_min_res_unit',  label  => 'Minimum reservation size',
                                                  info   => 'The minimum block size that is allocated for each entry.',
                                                  colour => '000000'},
            {name => 'avg_block_size',            label  => 'Average block size',
                                                  info   => 'The average size of currently allocated Qcache blocks.',
                                                  colour => '00CC00',
                                                  value  => sub {
                                                                  if ( $_[0]->{Qcache_queries_in_cache} > 0 ) {
                                                                    return (($_[0]->{QCache_total_blocks} - $_[0]->{Qcache_free_blocks}) * $_[0]->{query_cache_min_res_unit}) / $_[0]->{Qcache_queries_in_cache}
                                                                  } else {
                                                                    return 0
                                                                  }
                                                              }},
        ],
    },
}}

1;


contrib/Munin/MySQL/Graph/SchemaSize.pm
package Munin::MySQL::Graph::SchemaSize;

# Author <vegivamp@tuxera.be>
#
# Grabs the data_length and index_length grouped per schema.
#

sub collect_data {
    my $self = shift;
    my ($dbh) = @_;
    my $data = {};

    # This is a DIRTY HACK. Only way I could find to optimize the information_schema query, though.
    # Have a good gander at http://dev.mysql.com/doc/refman/5.1/en/innodb-parameters.html#sysvar_innodb_stats_on_metadata
    # It shouldn't have any impact due to only being flipped for a split second, but you never know.
    my $sth = $dbh->prepare('/*!50117 SELECT @@innodb_stats_on_metadata*/');
    $sth->execute;
    my ($innodb_stats_on_metadata) = @{$sth->fetch};
    $dbh->do('/*!50117 SET GLOBAL innodb_stats_on_metadata=0 */');

    my $query = 'SELECT table_schema, sum(data_length) as data_size, sum(index_length) as index_size'
              . '  FROM information_schema.tables'
              . '  GROUP BY table_schema';

    $sth = $dbh->prepare($query);
    $sth->execute();
    while (my $row = $sth->fetchrow_hashref) {
        $data->{'database_size'}->{$row->{'table_schema'}} = { 'data'  => $row->{'data_size'},
                                                               'index' => $row->{'index_size'}};
        $data->{'datasize_'  . $row->{'table_schema'}} = $row->{'data_size'};
        $data->{'indexsize_' . $row->{'table_schema'}} = $row->{'index_size'};
    }

    $dbh->do("/*!50117 SET GLOBAL innodb_stats_on_metadata=$innodb_stats_on_metadata */");
    $sth->finish();

    return $data;
}


sub graphs { 
    my $self = shift;
    my ($data) = @_;
    $graph = {
        schema_size => {
            config => {
                global_attrs => {
                    title => 'Schema data- and index size',
                    vlabel => 'index/data size in byte',
                    args => '--base 1024',

                },
                data_source_attrs => {
                    draw => 'LINE1',
                    type => 'GAUGE',
                    min  => 'U',
                },
            },
            data_sources => [],
        }
    };

    foreach $schema (keys %{$data->{'database_size'}}) {
      if ( $schema eq 'information_schema' or
           $schema eq 'performance_schema' ) {
        next;
      }
      push(@{$graph->{'schema_size'}->{'data_sources'}}, ({'name'     => 'indexsize_' . $schema,
                                                           'label'    => 'indexsize_' . $schema,
                                                           'graph'    => 'no'},
                                                          {'name'     => 'datasize_'  . $schema,
                                                           'label'    => $schema,
                                                           'negative' => 'indexsize_' . $schema}));
    }
    return $graph;
}

1;
